### PR TITLE
kubevirt basic test and specs

### PR DESCRIPTION
**Pull request by Rohit-PX**
**What type of PR is this?**
integration-test
Successful run: https://jenkins.pwx.dev.purestorage.com/job/Users/job/neelesh/job/stork-kubevirt-ocp/31/consoleFull

**What this PR does / why we need it**:

**Does this PR change a user-facing CRD or CLI?**:
no

**Is a release note needed?**:
no

**Does this change need to be cherry-picked to a release branch?**:
no

### CODE REVIEW

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+    storageclass.kubernetes.io/is-default-class: "true"
+  name: sc-sharedv4svc-nolock-remote-only
+parameters:
+  repl: "3"
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml:8]: Might be better to use "repl=2" here since it is more predicable for the tests especially when using multiple volumes. Or may be create multiple storage classes.

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  sharedv4: "true"
+  sharedv4_svc_type: ClusterIP
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml:10]: sharedv4_svc_type is not needed since "ClusterIP" is the default.

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  stork.libopenstorage.org/preferRemoteNodeOnly: "true"
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml:11]: can this be removed since we are ignoring it in the new implementation?

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  sharedv4_mount_options: vers=3.0,nolock
+provisioner: kubernetes.io/portworx-volume
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml:13]: Can we use CSI since that is what most new customers will be using?

File: test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      terminationGracePeriodSeconds: 180
+      volumes:
+      - dataVolume:
+          name: fedora-datavolume-clone
+        name: rootdisk
+      - cloudInitNoCloud:
+          userData: |-
+            #cloud-config
+            user: fedora
+            password: pum8-qftb-0y61
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml:65]: can we use some default password e.g. "fedora" in case the test wants to ssh into the VM?

File: test/integration_test/specs/kubevirt-fedora-vm/datavolume-clone.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+kind: DataVolume
+metadata:
+  name: fedora-datavolume-clone
+  annotations:
+    cdi.kubevirt.io/storage.bind.immediate.requested: "true"
+spec:
+  source:
+    pvc:
+      name: fedora-datavolume-original 
+      namespace: kubevirt-kubevirt-fedora-vm
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/datavolume-clone.yaml:11]: why "kubevirt-kubevirt"?

File: test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml [hunk 7]
```diff
@@ -0,0 +1,67 @@
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  annotations:
+    kubevirt.io/latest-observed-api-version: v1
+    kubevirt.io/storage-observed-api-version: v1alpha3
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml:6]: remove the annotations since they will be added by the kubevirt?

same for "generation"?

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+    storageclass.kubernetes.io/is-default-class: "true"
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml:5]: remove the "is-default-class" since it may conflict with other SCs?

File: Makefile [hunk 9]
```diff
@@ -57,12 +57,14 @@ all: stork storkctl cmdexecutor pretest px-statfs
 
 vendor-tidy:
 	go mod tidy
+	sed -i '1 i\// +build skipcompile\n' vendor/kubevirt.io/client-go/kubecli/kubevirt_test_utils.go
```
Comment (pureneelesh) [inline Makefile:60]: why is this needed in "go mod tidy"? Should be enough to add it to "vendor"?

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml (hunk 1 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml]: fixed.

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml (hunk 3 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml]: Fixed.

File: test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml (hunk 5 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml]: Done.

File: test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml (hunk 7 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora-vm/fedora-vm.yaml]: Fixed.

File: Makefile (hunk 9 above)
Comment (Rohit-PX) [reply Makefile]: Removed from `go mod tidy`

File: test/integration_test/specs/kubevirt-fedora-vm/datavolume-clone.yaml (hunk 6 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora-vm/datavolume-clone.yaml]: Fixed.

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml (hunk 8 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml]: SC is not default now.

**Review by pp511 — COMMENTED** [review]
Current changes look ok to me. 
Will wait for your final changes to handle pvc naming in the vm spec.

File: test/integration_test/kubevirt_test.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+	err = schedulerDriver.WaitForRunning(ctxs[0], defaultWaitTimeout, defaultWaitInterval)
+	require.NoError(t, err, "Error waiting for app to get to running state")
+
+	namespace := appKey + "-" + instanceID
+	vms, err := kubevirt.Instance().ListVirtualMachines(namespace)
+	require.NoError(t, err, "Error listing virtual machines")
+
+	for _, vm := range vms.Items {
+		validateVm(t, vm.Name, vm.Namespace, deployedVMName)
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:65]: nit: validateVm -> validateVM
why not pass the vm obj to validateVM() so that we don't have to call GetVirtualMachine() again?

File: test/integration_test/kubevirt_test.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	return ctxs
+}
+
+func validateVm(t *testing.T, name, namespace, vmName string) {
+	fedoraVm, err := kubevirt.Instance().GetVirtualMachine(name, namespace)
+	fmt.Printf("Fedora VM Name: %s", fedoraVm.Name)
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:72]: Where does this message appear? Should we use logrus or t.Logf() instead of fmt.Printf? Also, do you need "\n" in the end?

File: test/integration_test/kubevirt_test.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	require.NoError(t, err, "error in getting deployed VM in namespace: %s", namespace)
+	require.Equal(t, fedoraVm.Name, vmName, "VM %s has not been deployed in namespace: %s", vmName, namespace)
+
+	// TODO add more validations here if required
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:76]: Can we at least validate that the VM's status.created and status.Ready are true?
```    
    created: true
    ready: true
 ```

File: test/integration_test/specs/kubevirt-fedora/fedora-datavolume-template.yaml [hunk 13]
```diff
@@ -0,0 +1,18 @@
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataVolume
+metadata:
+  name: fedora-template-pvc
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/fedora-datavolume-template.yaml:4]: Do we need "-pvc" in the name? It might be confusing since this is not a pvc object. We could use "fedora-golden-image" or something like that.

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml [hunk 14]
```diff
@@ -0,0 +1,84 @@
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  #  annotations:
+  #    kubevirt.io/latest-observed-api-version: v1
+  #    kubevirt.io/storage-observed-api-version: v1alpha3
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:6]: remove?

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+  labels:
+    app: fedora-vm
+    vm.kubevirt.io/template: fedora-server-small
+    vm.kubevirt.io/template.namespace: openshift
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:10]: Just curious. Are all these labels required or will they be added automatically by kubevirt?

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+  name: test-vm-csi
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:11]: is the namespace required here? I see that the code uses a different namespace:
```
namespace := appKey + "-" + instanceID
```

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        vm.kubevirt.io/flavor: small
+        vm.kubevirt.io/os: fedora
+        vm.kubevirt.io/workload: server
+      creationTimestamp: null
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:20]: remove creationTimestamp? No need to supply null.

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+      labels:
+        kubevirt.io/domain: fedora-datavolume
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:22]: is this label correct? In my running VM, the domain is same as the name of the VM.
```
    template:
      metadata:
        labels:
          kubevirt.io/domain: fedora-communist-toucan
          kubevirt.io/size: small
 ```

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        pvc:
+          name: fedora-template-pvc
+          namespace: openshift-virtualization-os-images
+      pvc:
+        accessModes:
+        - ReadWriteMany
+        resources:
+          requests:
+            storage: 40Gi
+        storageClassName: sc-sharedv4svc-nolock-remote-only
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:83]: wrong storage class? Should be sc-sharedv4svc-nolock?

File: test/integration_test/kubevirt_test.go (hunk 10 above)
Comment (Rohit-PX) [reply test/integration_test/kubevirt_test.go:83]: fixed: `validateVm` ==>`validateVM`

We are calling `GetVirtualMachine()` only once.

File: test/integration_test/kubevirt_test.go (hunk 11 above)
Comment (Rohit-PX) [reply test/integration_test/kubevirt_test.go:90]: Removed the message.

File: test/integration_test/kubevirt_test.go (hunk 12 above)
Comment (Rohit-PX) [reply test/integration_test/kubevirt_test.go:94]: Fixed.

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml (hunk 15 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:10]: I think we add them, either via the UI or specs.

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml (hunk 19 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml]: I changed the name of the storage class.

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml (hunk 18 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:22]: Removed

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml (hunk 17 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:20]: Removed.

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml (hunk 16 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml:11]: Removed.

**Comment by Rohit-PX** [comment]
```
--- PASS: TestKubevirt (1164.47s)
    --- PASS: TestKubevirt/kubevirtDeployFedoraVMWithClonePVC (220.21s)
PASS

DONE 2 tests in 1247.384s
Tests passed
```

File: test/integration_test/kubevirt_test.go [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+	require.NoError(t, err, "Error listing virtual machines")
+
+	for _, vm := range vms.Items {
+		validateVM(t, vm.Name, vm.Namespace, deployedVMName)
+	}
+	return ctxs
+}
+
+func validateVM(t *testing.T, name, namespace, vmName string) {
+	currentVM, err := kubevirt.Instance().GetVirtualMachine(name, namespace)
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:90]: why is GetVirtualMachine() needed? We just called ListVirtualMachines() right before the call to this function. Can we not use the same VM object?

File: test/integration_test/kubevirt_test.go [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+	// TODO: wait for imported pod to complete
+	var importerPod *corev1.Pod
+	waitForImporterPodStart := func() (interface{}, bool, error) {
+		allPods, err := core.Instance().GetPods("openshift-virtualization-os-images", nil)
+		if err != nil || len(allPods.Items) == 0 {
+			return nil, true, fmt.Errorf("No importer pod not started yet. Retrying")
+		}
+		for _, pod := range allPods.Items {
+			if strings.Contains(pod.Name, importerPodPrefix) {
+				importerPod, err = core.Instance().GetPodByName(pod.Name, "openshift-virtualization-os-images")
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:117]: why is GetPodByName() needed? We have just called GetPods() in the same function. Why can't we use the pod in the allPods.Items?

File: test/integration_test/kubevirt_test.go [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+				logrus.Infof("Importer pod found: %s. Status: %s", importerPod.Name, importerPod.Status.Phase)
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:118]: this line should be removed since we already have it after the nil checks below

File: test/integration_test/kubevirt_test.go [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+func createTemplatePVC() error {
+	_, err := schedulerDriver.Schedule("",
+		scheduler.ScheduleOptions{
+			AppKeys:   []string{"kubevirt-templates"},
+			Namespace: "openshift-virtualization-os-images",
+		})
+	if err != nil {
+		return fmt.Errorf("error deploying kubevirt templates")
+	}
+	// TODO: wait for imported pod to complete
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:108]: remove this TODO? we are already doing this

File: test/integration_test/kubevirt_test.go [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+
+	// Importer pod has begun, now wait for it to get completed and to disappear
+	waitForImporterPodCompletion := func() (interface{}, bool, error) {
+		allPods, err := core.Instance().GetPods("openshift-virtualization-os-images", nil)
+		if err != nil {
+			return "", false, nil
+		}
+		for _, pod := range allPods.Items {
+			if strings.Contains(pod.Name, importerPodPrefix) {
+				importerPod, err = core.Instance().GetPodByName(pod.Name, "openshift-virtualization-os-images")
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:141]: GetPodByName not needed?

File: test/integration_test/kubevirt_test.go [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+				logrus.Infof("Importer pod found but waiting for completion: %s. Status: %s", importerPod.Name, importerPod.Status.Phase)
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:142]: this log line should be after err check

File: test/integration_test/kubevirt_test.go [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+				if err != nil {
+					return nil, false, fmt.Errorf("error getting importer pod")
+				}
+				if importerPod != nil && importerPod.Status.Phase == corev1.PodRunning {
+					return "", true, fmt.Errorf("importer pod %s found but not completed yet. Status: %s", importerPod.Name, importerPod.Status.Phase)
+				} else {
+					logrus.Infof("Importer pod: %s completed successfully", importerPod.Name)
```
Comment (pureneelesh) [inline test/integration_test/kubevirt_test.go:149]: we should check the phase to make sure that it was successful?

File: test/integration_test/specs/kubevirt-fedora/portworx/sc-sharedv4svc-nolock.yaml [hunk 27]
```diff
@@ -0,0 +1,11 @@
+apiVersion: storage.k8s.io/v1
```
Comment (pureneelesh) [inline test/integration_test/specs/kubevirt-fedora/portworx/sc-sharedv4svc-nolock.yaml:1]: should we remove this one? Why maintain 2 copies of the same thing? It could be confusing in future when you make changes. You have to change the other one not this one.

**Review by pureneelesh — APPROVED** [review]

File: test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml (hunk 4 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora-vm/portworx/sc-sharedv4svc-nolock-remote-only.yaml]: This will be switched to CSI at run time via a cli param to stork integration test.

File: test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml (hunk 14 above)
Comment (Rohit-PX) [reply test/integration_test/specs/kubevirt-fedora/fedora-vm.yaml]: removed.

File: Makefile [hunk 28]
```diff
@@ -63,6 +63,7 @@ vendor-update:
 
 vendor:
 	go mod vendor
+	sed -i '1 i\// +build skipcompile\n' vendor/kubevirt.io/client-go/kubecli/kubevirt_test_utils.go
```
Comment (pp511) [inline Makefile:66]: We should let folks know on Stork dev channel that go mod vendor will not work on Stork repo anymore. When will need to do make vendor. Else many people will end up wasting time to figure this out.

File: test/integration_test/kubevirt_test.go [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+)
+
+var templatePVCSpecs = map[string]string{
+	"fedora": "kubevirt-templates",
+}
+
+const (
+	importerPodPrefix            = "importer"
+	importerPodStartTimeOut      = 2 * time.Minute
+	importerPodCompletionTimeout = 20 * time.Minute
```
Comment (pp511) [inline test/integration_test/kubevirt_test.go:29]: nit...Timeout vs TimeOut. Chose one.

File: test/integration_test/kubevirt_test.go [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+func createTemplatePVC() error {
+	_, err := schedulerDriver.Schedule("",
+		scheduler.ScheduleOptions{
+			AppKeys:   []string{"kubevirt-templates"},
+			Namespace: "openshift-virtualization-os-images",
+		})
+	if err != nil {
+		return fmt.Errorf("error deploying kubevirt templates")
+	}
+	// TODO: wait for imported pod to complete
```
Comment (pp511) [inline test/integration_test/kubevirt_test.go:107]: nit..TODO needs to be removed.

File: test/integration_test/kubevirt_test.go [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+	require.Equal(t, virtualMachine.Status.Ready, true, "VM %s ready status is: %t", virtualMachine.Name, virtualMachine.Status.Ready)
+
+	// TODO add more validations here if required
+}
+
+func createTemplatePVC() error {
+	_, err := schedulerDriver.Schedule("",
+		scheduler.ScheduleOptions{
+			AppKeys:   []string{"kubevirt-templates"},
+			Namespace: "openshift-virtualization-os-images",
```
Comment (pp511) [inline test/integration_test/kubevirt_test.go:102]: nit...let's use a var for namespace as this is being used in multiple places in this function

File: test/integration_test/kubevirt_test.go [hunk 32]
```diff
@@ -0,0 +1,149 @@
...
+		})
+	if err != nil {
+		return fmt.Errorf("error deploying kubevirt templates")
+	}
+	// TODO: wait for imported pod to complete
+	waitForImporterPodStart := func() (interface{}, bool, error) {
+		allPods, err := core.Instance().GetPods("openshift-virtualization-os-images", nil)
+		if err != nil || len(allPods.Items) == 0 {
+			return nil, true, fmt.Errorf("No importer pod not started yet. Retrying")
```
Comment (pp511) [inline test/integration_test/kubevirt_test.go:111]: nit..returned error need not start with a capital letter

**Review by pp511 — APPROVED** [review]
LGTM. Just some nits.

File: test/integration_test/kubevirt_test.go (hunk 29 above)
Comment (Rohit-PX) [reply test/integration_test/kubevirt_test.go:29]: fixed, chose `Timeout`

File: test/integration_test/kubevirt_test.go (hunk 31 above)
Comment (Rohit-PX) [reply test/integration_test/kubevirt_test.go:102]: Fixed. Defined a variable `kubevirtTemplateNamespace`

File: test/integration_test/kubevirt_test.go (hunk 30 above)
Comment (Rohit-PX) [reply test/integration_test/kubevirt_test.go]: fixed.

File: Makefile (hunk 28 above)
Comment (Rohit-PX) [reply Makefile:66]: Agree, we always recommend using the `make ....` command to do anything, and hence everything is in the Makefile. We can broadcast this on the stork channel.
